pkg/apis/elasticsearch/v1/elasticsearch_types.go

```diff
@@ -33,7 +33,9 @@ type ElasticsearchSpec struct {
 
 	// NodeSets allow specifying groups of Elasticsearch nodes sharing the same configuration and Pod templates.
 	// +kubebuilder:validation:MinItems=1
-	NodeSets []NodeSet `json:"nodeSets"`
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	NodeSets []NodeSet `json:"nodeSets" patchStrategy:"merge" patchMergeKey:"name"`
```

Comment:
Hi @wmellouli. Could you help me to understand how those 2 keys help kustomize to use a given strategy instead of another ? I'm asking because it does not seem to change the CRD, so I don't understand how it is changing the way a resource is patched.

Also could you add some steps to reproduce your issue, maybe editing or adding them as a comment in #3878 ?

Thanks

Reply:
Hi @barkbay, 

`patchMergeKey` and `patchStrategy` are not `kubebuiler CRD Validation markers`: https://book.kubebuilder.io/reference/markers/crd-validation.html

=> so you will not see changes on CRD.

They are `go field tag`, that will be used later for example in `kubernetes apimachinery`: https://github.com/kubernetes/apimachinery/blob/master/third_party/forked/golang/json/fields.go#L62-L63

I admit that I can't find a lot of documentation on it. This is what I find: https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/#notes-on-the-strategic-merge-patch

```
...
In some cases, the list is replaced, not merged.

With a strategic merge patch, a list is either replaced or merged depending on its patch strategy. The patch strategy is specified by the value of the patchStrategy key in a field tag in the Kubernetes source code.
```

I detailed an example in https://github.com/elastic/cloud-on-k8s/issues/3878, to reproduce this problem (what I expect and what I get). 

Here a more detailed example. 

Given this files hierarchy:

```
eck-test-patch
|
|___ base
|    |___ kustomization.yaml    
|    |___ elasticsearch.yaml
|
|___ overlays
     |___ production
          |___  kustomization.yaml  
          |___  production-config.yaml 
```

content of files:

**eck-test-patch/base/kustomization.yaml**:

```
resources:
- elasticsearch.yaml

```

**eck-test-patch/base/elasticsearch.yaml**:

```
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: elasticsearch
spec:
  version: 7.9.2
  image: elasticsearch
  nodeSets:
    - name: masters
      count: 1
      config:
        node.master: true
        node.data: false
        node.ingest: false
    - name: ingest-data
      count: 3
      config:
        node.master: false
        node.data: true
        node.ingest: true
```

**eck-test-patch/overlays/production/kustomization.yaml**:

```
resources:
  - ../../base

patchesStrategicMerge:
  - production-config.yaml

```

**eck-test-patch/overlays/production/production-config.yaml**:


```
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: elasticsearch
spec:
  image: elasticsearch/elasticsearch:7.9.2
  nodeSets:
  - name: masters
    count: 3
  - name: ingest-data
    count: 6


```


When I apply this command `kustomize build .` at `eck-test-patch/overlays/production`, I was expecting a `merge` behavior for the `NodeSets` list, like that:

```
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: elasticsearch
spec:
  version: 7.9.2
  image: elasticsearch/elasticsearch:7.9.2
  nodeSets:
    - name: masters
      count: 3
      config:
        node.master: true
        node.data: false
        node.ingest: false
    - name: ingest-data
      count: 6
      config:
        node.master: false
        node.data: true
        node.ingest: true

```

But what I get, is a `replace` behavior:

```
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: elasticsearch
spec:
  image: elasticsearch/elasticsearch:7.9.2
  nodeSets:
  - count: 3
    name: masters
  - count: 6
    name: ingest-data
  version: 7.9.2

```

kustomize version: `v3.5.4`

Reply:
I'm sorry, I still don't understand exactly how the tag in the ECK binary is taken into account by `kustomize`. I tried to build and deploy ECK from your master branch:

```golang
	// NodeSets allow specifying groups of Elasticsearch nodes sharing the same configuration and Pod templates.
	// +kubebuilder:validation:MinItems=1
	// +patchMergeKey=name
	// +patchStrategy=merge
	NodeSets []NodeSet `json:"nodeSets" patchStrategy:"merge" patchMergeKey:"name"`
```
 
Then used the example you provided and still get the same result:

```yaml
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: elasticsearch
spec:
  image: elasticsearch/elasticsearch:7.9.2
  nodeSets:
  - count: 3
    name: masters
  - count: 6
    name: ingest-data
  version: 7.9.2
```

Reading [kustomize](https://kubernetes-sigs.github.io/kustomize/api-reference/glossary/#patchstrategicmerge) documentation : 
> Note that for custom resources, SMPs are treated as json merge patches.

See also this [comment](https://github.com/kubernetes-sigs/kustomize/issues/742#issuecomment-458650435): 
> patchesStrategicMerge treats the patch as a strategic merge patch when corresponding resource is native in Kubernetes. If a kind is not found, such as a custom resource, the patch is treated as json merge patch

My understanding is that CRDs support SMP but not custom ressources.
Let me know if I'm missing something obvious here.